### ISSUE TITLE
API command `sync` doesn't identify the `schema` correctly

### ISSUE THREAD

**Issue by phmalek** [glotzerlab/signac#230]
### Description

calling `sync` to synchronize dst project with src, a warning arises, that the schemas differ, although they seem to be the same (statepoint files are attached
[signac_statepoints_src.txt](https://github.com/glotzerlab/signac/files/3641944/signac_statepoints_src.txt)
[signac_statepoints_dst.txt](https://github.com/glotzerlab/signac/files/3641949/signac_statepoints_dst.txt)

). 

### To reproduce

`sync src -v -u`

### Error output

```
WARNING: The detected schemas of the two projects differ! Use --force to ignore.
Keys found only in the source schema: dt, totalSteps, k_bending, period, phi, seed, a, reversalRate, totalMonomers, f_p
Keys found only in the destination schema: dt, totalSteps, k_bending, period, phi, seed, a, reversalRate, totalMonomers, f_p
```

### System configuration

Please complete the following information:

 - Operating System [e.g. macOS]:
 - Version of Python [e.g. 3.7]:
 - Version of signac [e.g. 1.0]:

Or copy & paste the output of: `python -c 'import platform; print(platform.platform()); import sys; print(sys.version); import signac; print(signac.__version__)'`

**Comment by csadorf** [comment]
@phmalek Thank you for reporting! We'll try to have a look at this as soon as possible. In the meantime, you should be able to use `--force` to override this warning. Please let us know if this is a blocker for you work. Thank you!

**Comment by vishav1771** [comment]
Hi @csdorf @vyasr, I have gone through the `schema.py` for this. I think this behavior is due to
```
if not ignore_values:
            ret.update({k for k, v in self.items() if k in other and other[k] != v})
```
In this, we are comparing the value of `ProjectSchema`, which is dict of datatype as key and set of all value in jobs for that key as the value. I think for comparing the to the schema we should only compare the data type only not the set of value in jobs.

**Comment by csadorf** [comment]
@vishav1771 Assuming that I interpret the code correctly, then setting the `ignore_values` parameter to `True` means to ignore all values (and their types!) in the comparison. For example, assuming we compare the following two schemas:

```python
schema_a = {"foo": {int: (0, 1)}}
schema_b = {"foo": {int: (0, 2)}}
```
The schemas would be identical with `ignore_values=True`, but different with `ignore_values=False`.
Note, that the same would hold true for the following schemas:
```python
schema_a = {"foo": {int: (0, 1)}}
schema_b = {"foo": {float: (0., 2.)}}
```
In other words, for two schemas to be identical with `ignore_values=False` (the default), they have to have the same values and types.

Since this issue occurs non-deterministically, I assume it must have to do something with the order of the values in the tuples, so one time it is `{int: (0, 1)}` and the other time it is `{int: (1, 0)}`.

**Comment by vishav1771** [comment]
@csadorf  I don't think that is the case as we are storing these values in a set, not a tuple. I have a doubt regarding this code. While comparing the schema, for the sync shouldn't we pass `ignore_values=True` in `schema.difference` . 
```
if check_schema:
        schema_src = source.detect_schema()
        schema_dst = destination.detect_schema()
        if schema_dst and schema_src and schema_src != schema_dst:
            only_in_src = schema_src.difference(schema_dst)
            only_in_dst = schema_dst.difference(schema_src)
            if only_in_src or only_in_dst:
                raise SchemaSyncConflict(schema_src, schema_dst)
```

**Comment by csadorf** [comment]
@vishav1771 No, a synchronization would not be possible if the values differed, so that can't be it.

Btw, it would be easier for me to follow your arguments if you could copy & paste the link to the code snippet instead of copy & pasting the lines themselves, see the [help page](https://help.github.com/en/github/managing-your-work-on-github/creating-a-permanent-link-to-a-code-snippet) on how to do that.

**Comment by vishav1771** [comment]
@csadorf Does that  means the following project cann't be synchronized unless `check_schema=False`
```
for i in range(3):
    project_1.open_job({ 'a': i }).init()

for i in range(5):
    project_2.open_job({ 'a': i }).init()
```
And the statepoint given in the issue description have such type of condition that results into error.

**Comment by vyasr** [comment]
Sorry I've been swamped this week and haven't had much of a chance to look at this. I'll try and investigate a bit more tomorrow.

**Comment by vyasr** [comment]
@csadorf I just had a look at this issue, I think that @vishav1771 is correct. The issue is deterministic, and it's what Vishav pointed out: if you try to sync two projects where one project contains any jobs that are not present in the other, you will receive an error (because of the comparison triggered when `ignore_values=False`). I think that this behavior is unexpected, particularly because we advertise this feature as being similar to rsync, and a very common use-case is to create and work with new jobs on the remote and then synchronize them back to the local machine. If you look at the two schemas provided by @phmalek, he is basically creating new jobs in the src project and trying to have them copied to the dst workspace.

I don't use signac sync in my own work, so I had not realized that this was how synchronizing was defined. If your intended application was to only synchronize data between projects with _identical_ sets of jobs, I think that needs to be clarified in the documentation; however, I think that substantially limits the usefulness of the function. I'm curious if you think that there are significant reasons (data integrity, unintended copying, etc) that would result from allowing such a synchronization. If not, I think changing this behavior probably makes sense in some future version of signac, because I think it's very counterintuitive for users who expect source jobs that don't exist at the destination to simply be copied over. Alternatively, we could make the `check_schema` options a bit more fine-grained to allow only checking types but not values, and expose this to the CLI.

**Comment by csadorf** [comment]
@vyasr I see your point and I apparently misinterpreted the problem. I just ran a local test and it seems that indeed `ignore_values=True` might be the right solution.

@vishav1771 Could you
1. Implement a unit test that clearly illustrates and catches the problem (i.e. this unit test *should fail*), and then
2. implement the patch you suggested above (the unit test should now pass)?

**Comment by vishav1771** [comment]
@csadorf , I'll start working on this issue after I am done with pytest issue, which I am currently working on.

**Comment by vishav1771** [comment]
Should I compare the schema with data type instead of values or let the `detect_schema` as it is.

**Comment by csadorf** [comment]
> @csadorf , I'll start working on this issue after I am done with pytest issue, which I am currently working on.

OK!

> Should I compare the schema with data type instead of values or let the `detect_schema` as it is.

The best way to figure this out would be to implement one or multiple unit tests of different basic scenarios that clearly illustrate what would be the expected behavior. So I would just set up a few project combinations and then make an assessment whether it makes sense that they would be *syncable* or not. You gave an example above that should be fine, but does not work right now. You could create a few more examples and implement one unit test for each of them and check what the expected behavior would be. Here are a few scenarios I could come up with from the top of my head:
1. same keys, same types, same values
2. same keys, same types, different values
3. same keys, different types, same values
4. same keys, different types, different values
5. and so on

In essence, the correct solution to this problem should be guided by the tests.
